### PR TITLE
Add a wrapper for network configuration and load it when appropriate.

### PR DESCRIPTION
**Pull request by dmkozh**
# Description

Add a wrapper for network configuration and load it when appropriate.

Getting the right configuration from the ledger using transactions would be quite cumbersome. Instead, provide access to all the relevant entries via getters (probably some of them will not be relevant for core; we can clean this up if needed).

Also generalized the initialization and upgrades process, as now we have a lot of various settings.

This is the Core-side implementation of https://github.com/stellar/stellar-core/issues/3691

# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/ledger/NetworkConfig.cpp [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+createConfigSettingEntry(
+    ConfigSettingEntry const& configSetting, AbstractLedgerTxn& ltxRoot,
+    std::optional<uint32_t> mockLastModifiedLedgerSeqForTests = std::nullopt)
+{
+    LedgerEntry e;
+    e.data.type(CONFIG_SETTING);
+    e.data.configSetting() = configSetting;
+    if (mockLastModifiedLedgerSeqForTests)
+    {
+        e.lastModifiedLedgerSeq = *mockLastModifiedLedgerSeqForTests;
```
Comment (MonsieurNicolas) [inline src/ledger/NetworkConfig.cpp:24]: Not sure why you need this: see `LedgerTxn::Impl::commit()` that will write over `lastModifiedLedgerSeq` using `ledgerSeq`

**Review by jayz22 — APPROVED** [review]

File: src/ledger/LedgerManagerImpl.cpp [hunk 2]
```diff
@@ -222,7 +222,24 @@ LedgerManagerImpl::startNewLedger(LedgerHeader const& genesisLedger)
     SecretKey skey = SecretKey::fromSeed(mApp.getNetworkID());
 
     LedgerTxn ltx(mApp.getLedgerTxnRoot(), false);
+    auto const& cfg = mApp.getConfig();
+    // History catchup with BucketListDB enabled test fails if we put this here because last modified ledger seq is incorrect
```
[the remark this answers is not in the document]
Comment (dmkozh) [reply src/ledger/LedgerManagerImpl.cpp]: Thanks, I'm loading the configs lazily now in order to avoid these edge cases of v20 test initialization.

File: src/ledger/LedgerManagerImpl.cpp [hunk 3]
```diff
@@ -1064,6 +1089,27 @@ LedgerManagerImpl::advanceLedgerPointers(LedgerHeader const& header,
...
+    if (!upgradeHappened && mLastContractNetworkConfig)
+    {
+        return;
+    }
+    if (!mLastContractNetworkConfig)
+    {
+        mLastContractNetworkConfig =
+            std::make_optional<ContractNetworkConfig>();
+    }
+    LedgerTxn ltx(rootLtx, false, TransactionMode::READ_ONLY_WITHOUT_SQL_TXN);
```
Comment (sisuresh) [inline src/ledger/LedgerManagerImpl.cpp:1105]: This line is making me think that the LedgerTxns added in the [config upgrade PR](https://github.com/stellar/stellar-core/pull/3674/files) should be using `READ_ONLY_WITHOUT_SQL_TXN` since they should all be read only. I'm not sure yet if there are concurrency issues with this, but I forgot this option existed for `LedgerTxn` when I made those changes. What do you think?

1. https://github.com/stellar/stellar-core/blob/bff2c2b37402acc61166c0f369da1205efbe8c9d/src/herder/HerderImpl.cpp#L1210
2. https://github.com/stellar/stellar-core/blob/bff2c2b37402acc61166c0f369da1205efbe8c9d/src/herder/HerderImpl.cpp#L1278
3. https://github.com/stellar/stellar-core/blob/bff2c2b37402acc61166c0f369da1205efbe8c9d/src/herder/Upgrades.cpp#L632

File: src/ledger/NetworkConfig.h [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+{
+  public:
+    // Creates the initial contract configuration entries for protocol v20.
+    // This should happen once during the correspondent protocol version
+    // upgrade.
+    static void createLedgerEntriesForV20(AbstractLedgerTxn& ltx);
+    // Test-only function that initializes contract network configuration
+    // bypassing the normal upgrade process (i.e. when genesis ledger starts not
+    // at v1)
+    static void
```
Comment (sisuresh) [inline src/ledger/NetworkConfig.h:74]: This should be wrapped in a `BUILD_TESTS` `ifdef` like https://github.com/stellar/stellar-core/blob/bff2c2b37402acc61166c0f369da1205efbe8c9d/src/bucket/BucketManagerImpl.h#L147-L155

File: src/herder/test/UpgradesTests.cpp [hunk 5]
```diff
@@ -1973,8 +2030,7 @@ TEST_CASE("configuration initialized in version upgrade", "[upgrades]")
         ltx.load(getMaxContractSizeKey()).current().data.configSetting();
     REQUIRE(maxContractSizeEntry.configSettingID() ==
             CONFIG_SETTING_CONTRACT_MAX_SIZE_BYTES);
-    REQUIRE(maxContractSizeEntry.contractMaxSizeBytes() ==
-            Upgrades::CONFIG_SETTING_CONTRACT_MAX_SIZE_BYTES_V20);
+    REQUIRE(maxContractSizeEntry.contractMaxSizeBytes() == 65536);
```
[the remark this answers is not in the document]
Comment (dmkozh) [reply src/herder/test/UpgradesTests.cpp:2035]: Fixed.

File: src/ledger/NetworkConfig.h (hunk 4 above)
Comment (dmkozh) [reply src/ledger/NetworkConfig.h:74]: `USE_CONFIG_FOR_GENESIS` is not conditioned by this macro (probably it should be), so I can't use it here as well. I suppose we support this for standalone runs of core with bumped protocol version.

File: src/ledger/LedgerManagerImpl.cpp (hunk 3 above)
Comment (dmkozh) [reply src/ledger/LedgerManagerImpl.cpp:1105]: Updated.

File: src/ledger/NetworkConfig.h (hunk 4 above)
Comment (sisuresh) [reply src/ledger/NetworkConfig.h:74]: `USE_CONFIG_FOR_GENESIS` is not conditioned behind any macros, so wouldn't it still work?

File: src/ledger/LedgerManagerImpl.cpp (hunk 3 above)
Comment (sisuresh) [reply src/ledger/LedgerManagerImpl.cpp:1105]: Is there a difference between passing in true vs false for `shouldUpdateLastModified` when `TransactionMode == READ_ONLY_WITHOUT_SQL_TXN`? I don't think there is, but we pass in true for the current use cases, while the fee config PR passes in false.

Comment (sisuresh) [reply src/ledger/LedgerManagerImpl.cpp:1105]: Also, I opened a PR related to using `READ_ONLY_WITHOUT_SQL_TXN` - https://github.com/stellar/stellar-core/issues/3716.

File: src/ledger/NetworkConfig.h (hunk 4 above)
Comment (dmkozh) [reply src/ledger/NetworkConfig.h:74]: I mean that we need to call this when `USE_CONFIG_FOR_GENESIS ` is set to true. I'm not entirely sure why is it not guarded by `BUILD_TESTS` but for now in order to make the code compile without `BUILD_TESTS` we can't use this condition for my function.

File: src/herder/test/UpgradesTests.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (sisuresh) [reply src/herder/test/UpgradesTests.cpp:2035]: I don't see the fix

[the remark this answers is not in the document]
Comment (dmkozh) [reply src/herder/test/UpgradesTests.cpp]: There apparently were two of these; now both should be updated.

File: src/ledger/LedgerManagerImpl.cpp [hunk 6]
```diff
@@ -1064,6 +1089,27 @@ LedgerManagerImpl::advanceLedgerPointers(LedgerHeader const& header,
     mLastClosedLedger.header = header;
 }
 
+void
+LedgerManagerImpl::maybeUpdateNetworkConfig(bool upgradeHappened,
```
Comment (sisuresh) [inline src/ledger/LedgerManagerImpl.cpp:1093]: I think this code should be behind `
#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION`

Comment (dmkozh) [reply src/ledger/LedgerManagerImpl.cpp:1093]: Since the code compiles fine, I'd prefer to leave this as is. This doesn't touch any XDR and is conditioned on the protocol version at runtime.

Comment (dmkozh) [reply src/ledger/LedgerManagerImpl.cpp:1093]: Actually, I've realized that I've removed the guard from soroban protocol version in this PR as well, so I returned the guard for the sake of consistency and added guard to this function (now if you access settings in vcurr you're guarantteed to crash on accessing nullopt, which is good for catching bugs)

File: src/herder/test/UpgradesTests.cpp [hunk 7]
```diff
@@ -828,7 +827,47 @@ TEST_CASE("config upgrades applied to ledger", "[upgrades]")
...
+            configEntry2.configSettingID(
+                CONFIG_SETTING_CONTRACT_HISTORICAL_DATA_V0);
+            configEntry2.contractHistoricalData().feeHistorical1KB = 555;
+            LedgerTxn ltx(app->getLedgerTxnRoot());
+            configUpgradeSet = makeConfigUpgradeSet(ltx, configUpgradeSetXdr);
+            ltx.commit();
+        }
+        executeUpgrade(*app, makeConfigUpgrade(*configUpgradeSet));
+        REQUIRE(sorobanConfig.feeRatePerInstructionsIncrement() == 111);
+        LedgerTxn ltx(app->getLedgerTxnRoot());
```
Comment (sisuresh) [inline src/herder/test/UpgradesTests.cpp:866]: Nit this can be removed

**Comment by sisuresh** [comment]
r+ 2aaf792e05047b185dc251cec950ce7b5ab226da

File: src/herder/test/UpgradesTests.cpp (hunk 7 above)
Comment (dmkozh) [reply src/herder/test/UpgradesTests.cpp]: Done.
